src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java

```diff
@@ -620,47 +620,52 @@ private synchronized String getHeadsAsString() {
-                    // It's okay to fail to log
+                // Ensure that the Run is marked as completed (failed) if it isn't already so it won't show as running
+                Queue.Executable ex = owner.getExecutable();
+                if (ex instanceof Run) {
+                    Result res = ((Run) ex).getResult();
+                    setResult(res != null ? res : Result.FAILURE);
                 }
             }
 
-            // Switch to fallback storage so we don't delete original node data
-            this.storageDir = (this.storageDir != null) ? this.storageDir+"-fallback" : "workflow-fallback";
-            this.storage = createStorage();  // Empty storage
-
-            // Clear out old start nodes and heads
-            this.startNodes = new Stack<BlockStartNode>();
-            FlowHead head = new FlowHead(this);
-            this.heads = new TreeMap<Integer, FlowHead>();
-            heads.put(head.getId(), head);
-            FlowStartNode start = new FlowStartNode(this, iotaStr());
-            head.newStartNode(start);
-
-            // Create end
-            FlowNode end = new FlowEndNode(this, iotaStr(), (FlowStartNode)startNodes.pop(), result, getCurrentHeads().toArray(new FlowNode[0]));
-            end.addAction(new ErrorAction(failureReason));
-            head.setNewHead(end);
-            saveOwner();
+            try {
+                programPromise = Futures.immediateFailedFuture(new IllegalStateException("Failed loading heads", failureReason));
+                LOGGER.log(Level.INFO, "Creating placeholder flownodes for execution: " + this);
+                if (this.owner != null) {
+                    try {
+                        owner.getListener().getLogger().println("Creating placeholder flownodes because failed loading originals.");
+                    } catch (Exception ex) {
+                        // It's okay to fail to log
+                    }
+                }
 
+                // Switch to fallback storage so we don't delete original node data
+                this.storageDir = (this.storageDir != null) ? this.storageDir + "-fallback" : "workflow-fallback";
+                this.storage = createStorage();  // Empty storage
+
+                // Clear out old start nodes and heads
+                this.startNodes = new Stack<BlockStartNode>();
+                FlowHead head = new FlowHead(this);
+                this.heads = new TreeMap<Integer, FlowHead>();
+                heads.put(head.getId(), head);
+                FlowStartNode start = new FlowStartNode(this, iotaStr());
+                head.newStartNode(start);
+
+                // Create end
+                FlowNode end = new FlowEndNode(this, iotaStr(), (FlowStartNode) startNodes.pop(), result, getCurrentHeads().toArray(new FlowNode[0]));
+                end.addAction(new ErrorAction(failureReason));
+                head.setNewHead(end);
+            } catch (Exception ex) {
+                throw ex;
```

Comment:
Uh, what is this `try`-`catch`-`throw` accomplishing? Just delete it!

Reply:
Main reason is to separate out the synchronization components, but also try to ensure the run save happens.

Reply:
Wait, no.  I'm going to blame jet lag and lack of sleep for this. You're right, nonsensical.